toolkit/tools/internal/safechroot/safechroot.go

```diff
@@ -441,50 +451,85 @@ func cleanupAllChroots() {
 	inChrootMutex.Lock()
 
 	// mount is only supported in regular pipeline
+	failedToUnmount := false
 	if buildpipeline.IsRegularBuild() {
 		// Cleanup chroots in LIFO order incase any are interdependent (e.g. nested safe chroots)
 		logger.Log.Info("Cleaning up all active chroots")
 		for i := len(activeChroots) - 1; i >= 0; i-- {
 			logger.Log.Infof("Cleaning up chroot (%s)", activeChroots[i].rootDir)
-			err := activeChroots[i].unmountAndRemove(leaveChrootOnDisk)
+			err := activeChroots[i].unmountAndRemove(leaveChrootOnDisk, isFinalAttempt)
 			// Perform best effort cleanup: unmount as many chroots as possible,
 			// even if one fails.
 			if err != nil {
 				logger.Log.Errorf("Failed to unmount chroot (%s)", activeChroots[i].rootDir)
+				failedToUnmount = true
 			}
 		}
 	}
 
-	logger.Log.Info("Cleanup finished")
+	if failedToUnmount {
+		logger.Log.Fatalf("Failed to unmount a chroot, manual unmount required")
```

Comment:
Why do we want to hard crash here?

Reply:
Is the "manual" unmount something the user has to do? If yes, echoing the dangling mount point and command to unmount it might be helpful for the user.

Reply:
I'm torn on this. I settled on "stop immediately, something is VERY wrong" here since this should now "never" fail. Any subsequent actions risk damaging the host system here (although in practice the tools will just yell and refuse to continue if we try to re-use the chroot).

Generally the user will want to call `make clean-<sub component>` to clean the chroot, but the tools can't really know this. I guess I could recommend the user call `make clean` but it feels a bit weird to have a library knowing about the makefile several layers below it.

Reply:
I'd say for errors with an external source, I'd still return `err`, because we don't have control over them and really predict how they may change in the future. For cases where the error can only be caused by a bug in our tool's logic, then I think we can crash (I'm referring to the internal logic of the dependency graph we've discussed a while ago).

Reply:
I'll be ok with switching this over to an err that bubbles out if we see it failing in practice, but I don't expect it to happen.

Reply:
Is the reason you want to hard crash here related to the fact that the `Close()` error isn't being handled by the caller?

Reply:
Yes, that's a bit part of it. We have an emergency handler to clean up the chroots, but it only runs if the caller uses the logger's exit routines (which... we don't really do...)